[windowsservicereceiver] Initial implementation

merdzikdominik opened the pull request:
Fully working windowsservicereceiver.

Performed testing by running:
- make test
- make lint

merdzikdominik committed:
Add files via upload

merdzikdominik committed:
Update factory_windows.go

merdzikdominik committed:
Update factory.go

merdzikdominik committed:
Update scraper.go

merdzikdominik committed:
Update winapi.go

merdzikdominik committed:
Update winservice.go

merdzikdominik committed:
Update README.md

merdzikdominik committed:
Update README.md

merdzikdominik committed:
Update builder-config.yaml

shalper2 commented:
Het @merdzikdominik thanks for taking a look at this issue and contributing to it! Could you please include tests for the code that you are contributing?

pjanotti commented on cmd/otelcontribcol/builder-config.yaml:
```diff
@@ -244,6 +244,7 @@ receivers:
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/snmpreceiver v0.134.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowsperfcountersreceiver v0.134.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowseventlogreceiver v0.134.0
+  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowsservicereceiver v0.134.0
```
We want to wait to make this change per @shalper2 comment about the stability of the receiver.

pjanotti commented on receiver/windowsservicereceiver/factory_others.go:
```diff
@@ -0,0 +1,48 @@
+
+var errUnsupportedOS = errors.New("windowsservicereceiver: supported only on Windows")
+
+func NewFactory() receiver.Factory {
```
At first I think this will be better if you have the same `NewFactory` at `factory.go` but then changing line 26 to `receiver.WithMetrics(createMetricsReceiver, metadata.MetricsStability)` and then implement the platform dependent `createMetricsReceiver` in `factory_windows.go` and `factory_others.go`. See `receiver/windowsperfcountersreceiver` for an example.

pjanotti commented on receiver/windowsservicereceiver/scraper.go:
```diff
@@ -1,48 +1,145 @@
 //go:build windows
 
-package windowsservicereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowsservicereceiver"
+package windowsservicereceiver
```
same nit: let's keep the import comment

pjanotti commented on receiver/windowsservicereceiver/scraper.go:
```diff
@@ -1,48 +1,145 @@
-//nolint:unused
-func (*windowsServiceScraper) start(context.Context, component.Host) (err error) {
-	return nil
+func mapStartTypeToAttr(st StartType, _ bool) metadata.AttributeStartupMode {
```
Since it is a private function the addition of the not used `bool` can be delayed until it is really needed.

pjanotti commented on receiver/windowsservicereceiver/winservice.go:
```diff
@@ -1,39 +1,68 @@
 
-//nolint:unused
-func (*winService) getStatus() error {
+func (ws *winService) getStatus() error {
```
To me `getStatus` sounds like is returning status and possibly an error, like `getService` - perhaps a different name like `updateStatus` will make the difference clear. Same thing to `getConfig` below.

pjanotti reviewed (commented):
Thanks for working on this @merdzikdominik!

merdzikdominik committed:
Update winservice.go

replaced "get" prefixes with "update" for method names accordingly to their logic

merdzikdominik committed:
Update scraper.go

(1) Errors are collected and returned collectively at the end (errsCh channel + multierr.Append).

(2) Scrape is parallel (errgroup + concurrency-limiting semaphore).

(3) Status is cast directly from svc.status.State with range validation.

(4) Removed an unnecessary, unused bool from mapStartTypeToAttr; start has a clean signature.

(5) Method names indicate the mutation: updateService, updateStatus, updateConfig.

merdzikdominik committed:
Update factory_windows.go

(1) Errors are collected and returned collectively at the end (errsCh channel + multierr.Append).

(2) Scrape is parallel (errgroup + concurrency-limiting semaphore).

(3) Status is cast directly from svc.status.State with range validation.

(4) Removed an unnecessary, unused bool from mapStartTypeToAttr; start has a clean signature.

(5) Method names indicate the mutation: updateService, updateStatus, updateConfig.

merdzikdominik committed:
Update factory_others.go

(1) Errors are collected and returned collectively at the end (errsCh channel + multierr.Append).

(2) Scrape is parallel (errgroup + concurrency-limiting semaphore).

(3) Status is cast directly from svc.status.State with range validation.

(4) Removed an unnecessary, unused bool from mapStartTypeToAttr; start has a clean signature.

(5) Method names indicate the mutation: updateService, updateStatus, updateConfig.

merdzikdominik committed:
Update factory.go

(1) Errors are collected and returned collectively at the end (errsCh channel + multierr.Append).

(2) Scrape is parallel (errgroup + concurrency-limiting semaphore).

(3) Status is cast directly from svc.status.State with range validation.

(4) Removed an unnecessary, unused bool from mapStartTypeToAttr; start has a clean signature.

(5) Method names indicate the mutation: updateService, updateStatus, updateConfig.

merdzikdominik committed:
Update README.md

updated code owners section

merdzikdominik committed:
Update README.md

returned development state

merdzikdominik committed:
Update go.mod

added golang.org/x/sync/errgroup

merdzikdominik committed:
Update go.sum

merdzikdominik committed:
Update factory.go

merdzikdominik committed:
Update factory_others.go

merdzikdominik committed:
Update scraper.go

merdzikdominik committed:
Update winapi.go

merdzikdominik committed:
Update winservice.go

pjanotti commented:
@atoulme #38619 is the tracking issue for this receiver, I've been in contact with @merdzikdominik so we can keep progress on this receiver. I'll review if all the steps to add a new receiver are being properly followed.

merdzikdominik committed:
removed goroutine from scraper.go

merdzikdominik committed:
adjusted builder-config.yaml

merdzikdominik committed:
Merge pull request #1 from merdzikdominik/merdzikdominik-windowsservicereceiver-codereview1-nongoroutine

Merdzikdominik windowsservicereceiver codereview1 nongoroutine

pjanotti committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into merdzikdominik-windowsservicereceiver

pjanotti committed:
Undo changes to builder-config.yaml

pjanotti reviewed (commented):
If we get the file organization and API cleaned up I think we can move ahead. The next steps will be to:

1. Add an integration test (select a server running on GH runners and check that we get the expected metrics)
2. Refactor `serviceManager` (likely remove it) so we can directly use the `mgr.Mgr` (see comment)

merdzikdominik committed:
codereview2

merdzikdominik committed:
builder-config.yaml fr

merdzikdominik committed:
Merge branch 'merdzikdominik-windowsservicereceiver' into merdzikdominik-windowsservicereceiver-codereview2

merdzikdominik committed:
Merge pull request #2 from merdzikdominik/merdzikdominik-windowsservicereceiver-codereview2

Merdzikdominik windowsservicereceiver codereview2

pjanotti committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into merdzikdominik-windowsservicereceiver

pjanotti committed:
make tidy

pjanotti committed:
Remove changes to build-config.yaml

pjanotti committed:
make generate

pjanotti committed:
Merge branch 'main' into merdzikdominik-windowsservicereceiver
